modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java

```diff
@@ -0,0 +1,449 @@
+        logger.debug("Processing database [{}] for configuration [{}]", name, id);
+
+        final String sha256Url = downloadUrl(name, "tar.gz.sha256");
+        final String tgzUrl = downloadUrl(name, "tar.gz");
+
+        final Pattern checksumPattern = Pattern.compile("(\\w{64})\\s\\s(.*)");
+        String result = new String(httpClient.getBytes(auth, sha256Url), StandardCharsets.UTF_8).trim(); // this throws if the auth is bad
+        var matcher = checksumPattern.matcher(result);
+        boolean match = matcher.matches();
+        if (match == false) {
+            throw new RuntimeException("Unexpected sha256 response from [" + sha256Url + "]");
+        }
+        final String sha256 = matcher.group(1);
+        // the name that comes from the enterprise downloader cluster state doesn't include the .mmdb extension,
+        // but the downloading and indexing of database code expects it to be there, so we add it on here before further processing
+        processDatabase(auth, name + ".mmdb", sha256, tgzUrl);
+    }
+
+    private void processDatabase(PasswordAuthentication auth, String name, String sha256, String url) {
+        Metadata metadata = state.getDatabases().getOrDefault(name, Metadata.EMPTY);
+        if (Objects.equals(metadata.sha256(), sha256)) {
+            updateTimestamp(name, metadata);
+            return;
+        }
+        logger.debug("downloading geoip database [{}]", name);
+        long start = System.currentTimeMillis();
+        try (InputStream is = httpClient.get(auth, url)) {
+            int firstChunk = metadata.lastChunk() + 1; // if there is no metadata, then Metadata.EMPTY + 1 = 0
+            Tuple<Integer, String> tuple = indexChunks(name, is, firstChunk, MessageDigests.sha256(), sha256, start);
+            int lastChunk = tuple.v1();
+            String md5 = tuple.v2();
+            if (lastChunk > firstChunk) {
+                state = state.put(name, new Metadata(start, firstChunk, lastChunk - 1, md5, start, sha256));
+                updateTaskState();
+                logger.info("successfully downloaded geoip database [{}]", name);
+                deleteOldChunks(name, firstChunk);
+            }
+        } catch (Exception e) {
+            logger.error(() -> "error downloading geoip database [" + name + "]", e);
+        }
+    }
+
+    // visible for testing
+    void deleteOldChunks(String name, int firstChunk) {
+        BoolQueryBuilder queryBuilder = new BoolQueryBuilder().filter(new MatchQueryBuilder("name", name))
+            .filter(new RangeQueryBuilder("chunk").to(firstChunk, false));
+        DeleteByQueryRequest request = new DeleteByQueryRequest();
+        request.indices(DATABASES_INDEX);
+        request.setQuery(queryBuilder);
+        client.execute(
+            DeleteByQueryAction.INSTANCE,
+            request,
+            ActionListener.wrap(r -> {}, e -> logger.warn("could not delete old chunks for geoip database [" + name + "]", e))
```

Comment:
If this deletion fails, we'll just retry it the next time around, right?

Reply:
Yes, that's something I really like about the existing logic in `GeoIpDownloader` -- because it deletes all old chunks for the name (every chunk less than the current `firstChunk`), it'll just retry on a later update of the same database if one of these should happen to fail. It's clever!

For that matter, the way it's implemented doesn't rely on chunks being deleted from the database ever -- it's just a GC process, besides eating up disk space there's no harm in keeping old chunks around forever (so to speak).